server/mqtt.go

```diff
@@ -2503,106 +2710,217 @@ func (sess *mqttSession) update(filters []*mqttFilter, add bool) error {
 	}
-	// Get the stream sequence and other from the ack reply subject
-	sseq, _, dcount := ackReplyInfo(reply)
 
-	var pending *mqttPending
-	// For this JS consumer, check to see if we already have sseq->pi
+	// Get the stream sequence and duplicate flag from the ack reply subject.
+	sseq, _, dcount := ackReplyInfo(jsAckSubject)
+	if dcount > 1 {
+		dup = true
+	}
+
+	var ack *mqttPending
 	sseqToPi, ok := sess.cpending[jsDur]
 	if !ok {
 		sseqToPi = make(map[uint64]uint16)
 		sess.cpending[jsDur] = sseqToPi
-	} else if pi, ok = sseqToPi[sseq]; ok {
-		// If we already have a pi, get the ack so we update it.
-		// We will reuse the save packet identifier (pi).
-		pending = sess.pending[pi]
+	} else {
+		pi = sseqToPi[sseq]
 	}
-	if pi == 0 {
+
+	if pi != 0 {
+		// There is a possible race between a PUBLISH re-delivery calling us,
+		// and a PUBREC received already having submitting a PUBREL into JS . If
+		// so, indicate no need for (re-)delivery by returning a PI of 0.
+		_, usedForPubRel := sess.pendingPubRel[pi]
+		if /*dup && */ usedForPubRel {
+			return 0, false
+		}
+
+		// We should have a pending JS ACK for this PI.
+		ack = sess.pendingPublish[pi]
+	} else {
 		// sess.maxp will always have a value > 0.
-		if len(sess.pending) >= int(sess.maxp) {
+		if len(sess.pendingPublish) >= int(sess.maxp) {
 			// Indicate that we did not assign a packet identifier.
 			// The caller will not send the message to the subscription
 			// and JS will redeliver later, based on consumer's AckWait.
 			return 0, false
 		}
-		pi = bumpPI()
+
+		pi = sess.bumpPI()
+		if pi == 0 {
+			return 0, false
+		}
+
 		sseqToPi[sseq] = pi
 	}
-	if pending == nil {
-		pending = &mqttPending{jsDur: jsDur, sseq: sseq, subject: reply}
-		sess.pending[pi] = pending
-	}
-	// If redelivery, return DUP flag
-	if dcount > 1 {
-		dup = true
+
+	if ack == nil {
+		sess.pendingPublish[pi] = &mqttPending{
+			jsDur:        jsDur,
+			sseq:         sseq,
+			jsAckSubject: jsAckSubject,
+		}
+	} else {
+		ack.jsAckSubject = jsAckSubject
+		ack.sseq = sseq
+		ack.jsDur = jsDur
 	}
+
 	return pi, dup
 }
 
-// Sends a request to create a JS Durable Consumer based on the given consumer's config.
-// This will wait in place for the reply from the server handling the requests.
+// Stops a PI from being tracked as a PUBLISH. It can still be in use for a
+// pending PUBREL.
 //
 // Lock held on entry
-func (sess *mqttSession) createConsumer(consConfig *ConsumerConfig) error {
-	cfg := &CreateConsumerRequest{
-		Stream: mqttStreamName,
-		Config: *consConfig,
+func (sess *mqttSession) untrackPublish(pi uint16) (jsAckSubject string) {
+	ack, ok := sess.pendingPublish[pi]
+	if !ok {
+		return _EMPTY_
 	}
-	_, err := sess.jsa.createConsumer(cfg)
-	return err
+
+	delete(sess.pendingPublish, pi)
+	if len(sess.pendingPublish) == 0 {
+		sess.last_pi = 0
+	}
+
+	if len(sess.cpending) != 0 && ack.jsDur != _EMPTY_ {
+		if sseqToPi := sess.cpending[ack.jsDur]; sseqToPi != nil {
```

Comment:
When using `map[x]map[y]z` you need to make sure that you clean up the top-level key if the second-level map is now `len(...) == 0`, otherwise we can end up with a map pointing to a bunch of empty maps.

Better approach if you are only dealing with base types is to define a composite key type struct:
```
type foo struct {
  bar string
  baz uint64
}
```
... and have a single-level `map[foo]qux`, so that you don't have to worry about nesting and top-level deletes.